[JSC] Optimize JSMap::Get for DFG and FTL by utilizing data pointer instead of data index

hyjorc1 opened the pull request:
#### b250c04ecf7f8f4c76db0ebf41bafe2886346e81
<pre>
[JSC] Optimize JSMap::Get for DFG and FTL by utilizing data pointer instead of data index
<a href="https://bugs.webkit.org/show_bug.cgi?id=276650">https://bugs.webkit.org/show_bug.cgi?id=276650</a>
<a href="https://rdar.apple.com/131813452">rdar://131813452</a>

Reviewed by Yusuke Suzuki.

JSMap is backed by a single array. Previously, JSMap::Get generates two
DFG nodes `MapKeyIndex` and `MapValue`, where:
* MapKeyIndex: Find the key index with a map storage and a key value.
* MapValue: Load the corresponding value with a map storage and a key index.

So, instead of loading and using the same map storage twice, we could do
* MapGet: Find the key slot in the map storage.
* MapValue: Load the value with the corresponding key slot.

Microbenchmark results with 20 iterations:

                                             without                     with

js-map-get-string                         5.6139+-0.0185     ^      5.5335+-0.0207        ^ definitely 1.0145x faster
js-map-get-string-no-dfg-no-inline       14.2257+-0.0283           14.1996+-0.0162
js-map-get-int                            5.7287+-0.0175     ^      5.5497+-0.0185        ^ definitely 1.0323x faster
js-map-get-int-no-dfg-no-inline          20.2157+-0.0181     ?     20.2332+-0.0168        ?

&lt;geometric&gt;                               9.8062+-0.0158     ^      9.6913+-0.0162        ^ definitely 1.0119x faster

* JSTests/microbenchmarks/js-map-get-int-no-dfg-no-inline.js:
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-int.js:
(testGet):
* JSTests/microbenchmarks/js-map-get-string-no-dfg-no-inline.js:
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-string.js:
(testGet):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileMapGetImpl):
(JSC::DFG::SpeculativeJIT::compileMapGet):
(JSC::DFG::SpeculativeJIT::compileLoadMapValue):
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::compileGetMapIndexImpl): Deleted.
(JSC::DFG::SpeculativeJIT::compileMapKeyIndex): Deleted.
(JSC::DFG::SpeculativeJIT::compileMapValue): Deleted.
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/OrderedHashTable.h:
(JSC::OrderedHashTable::getKeySlot):
(JSC::OrderedHashTable::getKeyIndex): Deleted.
* Source/JavaScriptCore/runtime/OrderedHashTableHelper.h:
(JSC::OrderedHashTableHelper::find):

Canonical link: <a href="https://commits.webkit.org/281039@main">https://commits.webkit.org/281039@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b3d8c1171042bce2c71f4900864c08e81db4551

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58428 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37755 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10912 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62053 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8872 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45391 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9069 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47326 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6334 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35365 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50519 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28171 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32120 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7818 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7876 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51519 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54072 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8088 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63757 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57670 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2341 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8120 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54645 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2349 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50541 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Running apply-patch; Checked out pull request; Skipped layout-tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54712 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1982 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79430 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8723 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33584 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13215 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34670 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35754 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34415 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -1586,6 +1589,7 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
         break;
 
     case MapIteratorNext:
+    case isEmptyStorage:
```
DFG node name should start with capital letter. So `IsEmptyStorage`.

Constellation commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -2719,12 +2719,12 @@ class FixupPhase : public Phase {
-            fixEdge<Int32Use>(node->child2());
+        case LoadMapValue:
+        case isEmptyStorage:
+            fixEdge<KnownCellUse>(node->child1());
```
Storage is not a cell. So it should just leave `Untyped`.

Constellation reviewed (approved):
r=me with comments.

hyjorc1 committed:
[JSC] Optimize JSMap::Get for DFG and FTL by utilizing data pointer instead of data index
https://bugs.webkit.org/show_bug.cgi?id=276650
rdar://131813452

Reviewed by Yusuke Suzuki.

JSMap is backed by a single array. Previously, JSMap::Get generates two
DFG nodes `MapKeyIndex` and `MapValue`, where:
* MapKeyIndex: Find the key index with a map storage and a key value.
* MapValue: Load the corresponding value with a map storage and a key index.

So, instead of loading and using the same map storage twice, we could do
* MapGet: Find the key slot in the map storage.
* MapValue: Load the value with the corresponding key slot.

Microbenchmark results with 20 iterations:

                                             without                     with

js-map-get-string                         5.6139+-0.0185     ^      5.5335+-0.0207        ^ definitely 1.0145x faster
js-map-get-string-no-dfg-no-inline       14.2257+-0.0283           14.1996+-0.0162
js-map-get-int                            5.7287+-0.0175     ^      5.5497+-0.0185        ^ definitely 1.0323x faster
js-map-get-int-no-dfg-no-inline          20.2157+-0.0181     ?     20.2332+-0.0168        ?

<geometric>                               9.8062+-0.0158     ^      9.6913+-0.0162        ^ definitely 1.0119x faster

* JSTests/microbenchmarks/js-map-get-int-no-dfg-no-inline.js:
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-int.js:
(testGet):
* JSTests/microbenchmarks/js-map-get-string-no-dfg-no-inline.js:
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-string.js:
(testGet):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileMapGetImpl):
(JSC::DFG::SpeculativeJIT::compileMapGet):
(JSC::DFG::SpeculativeJIT::compileLoadMapValue):
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::compileGetMapIndexImpl): Deleted.
(JSC::DFG::SpeculativeJIT::compileMapKeyIndex): Deleted.
(JSC::DFG::SpeculativeJIT::compileMapValue): Deleted.
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/OrderedHashTable.h:
(JSC::OrderedHashTable::getKeySlot):
(JSC::OrderedHashTable::getKeyIndex): Deleted.
* Source/JavaScriptCore/runtime/OrderedHashTableHelper.h:
(JSC::OrderedHashTableHelper::find):

Canonical link: https://commits.webkit.org/281039@main
